transcriptomics_data_service/routers/ingest.py

```diff
@@ -26,30 +26,72 @@ async def ingest(
-    # TODO read counts as integers
-
-    experiment_result = ExperimentResult(
-        experiment_result_id=experiment_result_id, assembly_name=assembly_name, assembly_id=assembly_id
-    )
-
+    rcm_df = _load_csv(file_bytes)
     # Perform the ingestion in a transaction
     async with db.transaction_connection() as transaction_con:
         # For each matrix: create ONE row in ExperimentResult
+        experiment_result = ExperimentResult(
+            experiment_result_id=experiment_result_id, assembly_name=assembly_name, assembly_id=assembly_id
+        )
         await db.create_experiment_result(experiment_result, transaction_con)
 
-        # TODO For EACH cell in the matrix: create one row in GeneExpression
-        # gene_expressions: list[GeneExpression]
-        # await db.create_gene_expressions(gene_expressions, transaction_con)
+        gene_expressions: list[GeneExpression] = []
+        for gene_code, row in rcm_df.iterrows():
+
+            for sample_id, raw_count in row.items():
+                gene_expression = GeneExpression(
+                    gene_code=gene_code,
+                    sample_id=sample_id,
+                    experiment_result_id=experiment_result_id,
+                    raw_count=raw_count,
+                )
+                gene_expressions.append(gene_expression)
+        await db.create_gene_expressions(gene_expressions, transaction_con)
     return
 
 
+def _load_csv(file_bytes: bytes) -> pd.DataFrame:
+    buffer = StringIO(file_bytes.decode("utf-8"))
+    buffer.seek(0)
+    try:
+        df = pd.read_csv(buffer, index_col=0, header=0)
+
+        # Check for duplicate values in index column or header
+        if df.index.duplicated().any():
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST, detail="Duplicate values found in the Gene ID column."
+            )
+        if df.columns.duplicated().any():
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST, detail="Duplicate values found in the Sample ID row."
+            )
+
+        try:
+            df = df.applymap(lambda x: int(x) if pd.notna(x) else x)
+        except ValueError as e:
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
+                detail=f"Data type error: Non-integer value found: {e}",
+            )
+
+        return df
+    except pd.errors.ParserError as e:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail=f"CSV parsing error: {e}",
+        )
+    except ValueError as e:
+        raise HTTPException(
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail=f"Data type error in CSV: {e}",
+        )
+    except Exception as e:
+        raise HTTPException(
+            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+            detail=f"Unexpected error while reading CSV: {e}",
+        )
```

Comment:
This catches the `HTTPExceptions` raised from `df.index.duplicated().any()` `df.columns.duplicated().any()`, effectively converting the original 400 status to a 500. We should remove this block, since any raised `Exception` that could occur here would result in a 500 response anyway.

Reply:
HTTPException is terminal, so if once raised cannot be caught by any other try/catch block. It terminates the thread after sending the response.